// @taskforcesh/bullmq-pro@7.26.1/dist/cjs/classes/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Worker = exports.Queue = exports.QueueEvents = exports.Job = exports.FlowProducer = exports.WaitingChildrenError = exports.UnrecoverableError = exports.getNextMillis = exports.Repeat = exports.RateLimitError = exports.RedisConnection = exports.QueueEventsProducer = exports.QueueBase = exports.JobScheduler = exports.DelayedError = exports.ChildProcessor = exports.ChildPool = void 0;
const tslib_1 = require("tslib");
var bullmq_1 = require("bullmq");
Object.defineProperty(exports, "ChildPool", { enumerable: true, get: function () { return bullmq_1.ChildPool; } });
Object.defineProperty(exports, "ChildProcessor", { enumerable: true, get: function () { return bullmq_1.ChildProcessor; } });
Object.defineProperty(exports, "DelayedError", { enumerable: true, get: function () { return bullmq_1.DelayedError; } });
Object.defineProperty(exports, "JobScheduler", { enumerable: true, get: function () { return bullmq_1.JobScheduler; } });
Object.defineProperty(exports, "QueueBase", { enumerable: true, get: function () { return bullmq_1.QueueBase; } });
Object.defineProperty(exports, "QueueEventsProducer", { enumerable: true, get: function () { return bullmq_1.QueueEventsProducer; } });
Object.defineProperty(exports, "RedisConnection", { enumerable: true, get: function () { return bullmq_1.RedisConnection; } });
Object.defineProperty(exports, "RateLimitError", { enumerable: true, get: function () { return bullmq_1.RateLimitError; } });
Object.defineProperty(exports, "Repeat", { enumerable: true, get: function () { return bullmq_1.Repeat; } });
Object.defineProperty(exports, "getNextMillis", { enumerable: true, get: function () { return bullmq_1.getNextMillis; } });
Object.defineProperty(exports, "UnrecoverableError", { enumerable: true, get: function () { return bullmq_1.UnrecoverableError; } });
Object.defineProperty(exports, "WaitingChildrenError", { enumerable: true, get: function () { return bullmq_1.WaitingChildrenError; } });
// aliases are used for testing bullmq base tests
var flow_producer_pro_1 = require("./flow-producer-pro");
Object.defineProperty(exports, "FlowProducer", { enumerable: true, get: function () { return flow_producer_pro_1.FlowProducerPro; } });
var job_pro_1 = require("./job-pro");
Object.defineProperty(exports, "Job", { enumerable: true, get: function () { return job_pro_1.JobPro; } });
var queue_events_pro_1 = require("./queue-events-pro");
Object.defineProperty(exports, "QueueEvents", { enumerable: true, get: function () { return queue_events_pro_1.QueueEventsPro; } });
var queue_pro_1 = require("./queue-pro");
Object.defineProperty(exports, "Queue", { enumerable: true, get: function () { return queue_pro_1.QueuePro; } });
var worker_pro_1 = require("./worker-pro");
Object.defineProperty(exports, "Worker", { enumerable: true, get: function () { return worker_pro_1.WorkerPro; } });
tslib_1.__exportStar(require("./flow-producer-pro"), exports);
tslib_1.__exportStar(require("./job-pro"), exports);
tslib_1.__exportStar(require("./queue-events-pro"), exports);
tslib_1.__exportStar(require("./queue-pro"), exports);
tslib_1.__exportStar(require("./worker-pro"), exports);
tslib_1.__exportStar(require("./errors"), exports);
